### PR TITLE
Responsys modifications

### PR DESCRIPTION
**Pull request by florrain**
## Summary

The [original repo](https://github.com/gzigzigzeo/sidekiq-grouping) where the code comes from is really great and almost fitted what I needed to make Responsys calls batched in a cleanish way.

I'll detail the original features then what I changed/added.
## A long time ago, in a repo not so far away

I'd define sidekiq-grouping as a Sidekiq middleware that acts as a filter before jobs get actually pushed to Sidekiq queues. If sidekiq_options are defined within the worker class and include sidekiq-grouping settings then it'll be considered a batch and the job will be deviated to a temporary queue in Redis, waiting for the right time to be flushed.

When all the criterias are filled, all the jobs that are in the same temp queue are submitted to Sidekiq in one single time. 

Here is an example of how _ORIGINALLY_ sidekiq-grouping used to work (I used the [README](https://github.com/gzigzigzeo/sidekiq-grouping/blob/master/README.md)):

``` ruby
# You define a Sidekiq Job
class SidekiqJobWorker
  include Sidekiq::Worker

  sidekiq_options(
    queue: :batched_jobs,
    batch_flush_size: 30,     # Jobs will be combined when queue size exceeds 30
    batch_flush_interval: 60 # Jobs will be combined every 60 seconds
  )

  def perform(batch_data)
  end
end
```

``` ruby
# You enqueue async jobs
SidekiqJobWorker.perform_async("arg1", "arg2")
SidekiqJobWorker.perform_async("arg3", "arg4")
SidekiqJobWorker.perform_async("arg5", "arg6")
```

``` ruby
# After a while, the perform method gets called:
def perform(batch_data)
    # Here you have access to the data.
    #
    # -> Only one parameter is allowed for the perform method when it's a s-g job.
    #
    # The batch_data comes in as follows:
    # [
    #   ["arg1", "arg2"],
    #   ["arg3", "arg4"],
    #   ["arg5", "arg6"]
    # ]
end
```
## Yes but how about Oracle Responsys API?

Responsys has an API throttle policy that limits the number of calls per minute for their Interact API. The rules are:
1 - The main factor is the number of calls per minute **allowed for that specific method**. It can be different for each method of the API. It can be 30, 100, 1000...
3 - The number of records per call: 200 maximum.

So let's make some changes to adapt the GEM based on these rules.
### Global settings (in a Rails initializer for example):

``` ruby
Sidekiq::Grouping::Config.enabled = true
Sidekiq::Grouping::Config.poll_interval = 3 # not changed
Sidekiq::Grouping::Config.batch_flush_interval = 60 # not changed
Sidekiq::Grouping::Config.max_records_per_call = 200 # default value
Sidekiq::Grouping::Config.max_calls_per_minute = 30 # default value
```
- poll_interval (sec): internal interval to check every single batch if flushable or not.
- enabled: if turned to false, the jobs are enqueued but not processed. (i.e for test or local environments)
- batch_flush_interval (sec): the interval a batch should be flushed.
### Sidekiq options

Options specific to a Job.

``` ruby
 sidekiq_options(
   grouping: true,
   queue: :responsys_batches, # not changed
   max_calls_per_minute: 150,
   max_records_per_call: 200
 )
```
- grouping: there was no flag that could be set for a batch job. The GEM was looking for its own specific settings and if one was set then it knew it was a batch. A boolean makes more sense for everybody.
- queue: nothing has changed here, it's the regular Sidekiq param that is used to move the job to the right queue.
- max_calls_per_minute and max_records_per_call are specific to Responsys and that particular Job.
### The input and output data is different

I needed to group jobs following the same pattern. I couldn't just group jobs of a same method together because an API call to Responsys acts most of the time on one Interact Object and even sometimes with different merge rules. 
So I changed the way the jobs are temporarily enqueued and added a _queue_option_ argument to all jobs.
It is the first parameter of a batched job, is mandatory and it needs to be a String. It's also returned in the perform data.
It will be used to group the jobs against the same queue option and the same job class.

**Example:**

``` ruby
# For these Sidekiq options
 sidekiq_options(
   grouping: true,
   queue: :responsys_batches,
   max_calls_per_minute: 3,
   max_records_per_call: 2
 )
```

``` ruby
# Input. Enqueue single modifications -> one job = one record to edit
MergeListMembersWorker.perform_async("CONTACTS_LIST,update_record", "user1@email.com", "1000") #1
MergeListMembersWorker.perform_async("CONTACTS_LIST,update_record", "user2@email.com", "2000") #2
MergeListMembersWorker.perform_async("CONTACTS_LIST,update_record", "user3@email.com", "3000") #3
MergeListMembersWorker.perform_async("CONTACTS_LIST,update_record", "user4@email.com", "4000") #4
MergeListMembersWorker.perform_async("CONTACTS_LIST,update_record", "user5@email.com", "5000") #5
MergeListMembersWorker.perform_async("CONTACTS_LIST,update_record", "user6@email.com", "6000") #6
MergeListMembersWorker.perform_async("CONTACTS_LIST,update_record", "user7@email.com", "7000") #7
```

``` ruby
# Output (#perform)
# t + 60s
#3 calls of 2 records = 6 records to update
{
  "queue_option" => "CONTACTS_LIST,update_record",
  "chunks" => [
    [
      ["CONTACTS_LIST,update_record", "user1@email.com", "1000"],
      ["CONTACTS_LIST,update_record", "user2@email.com", "2000"]
    ],
    [
      ["CONTACTS_LIST,update_record", "user3@email.com", "3000"],
      ["CONTACTS_LIST,update_record", "user4@email.com", "4000"]
    ],
    [
      ["CONTACTS_LIST,update_record", "user5@email.com", "5000"],
      ["CONTACTS_LIST,update_record", "user6@email.com", "6000"]
    ]
  ]
}

# t + 120s
#1 call of 1 records = 1 record to update
{
  "queue_option" => "option",
  "chunks" => [
    [
      ["CONTACTS_LIST,update_record", "user7@email.com", "7000"]
    ]
  ]
}
```

The Job is now able to handle the calls with the appropriate number of records.

I also updated the Web UI to add the queue option. To save real estate on the page, I grouped the workers per queue name. That also looks more visible to me if there is more than one queue.
## What's next?
- Add coverage on actor.rb which is were the number of records and calls are determined. This will be the occasion of cleaning up the code to make it easily to read and test.
- Update README

### CODE REVIEW

File: lib/sidekiq/grouping/actor.rb [hunk 1]
```diff
@@ -9,23 +9,53 @@ def initialize
       end
 
       private
-      
+
       def start_polling
+        return unless Sidekiq::Grouping::Config.enabled
```
Comment (dandemeyere) [inline lib/sidekiq/grouping/actor.rb:14]: Out of curiosity, why is this needed?

File: lib/sidekiq/grouping/actor.rb [hunk 2]
```diff
@@ -9,23 +9,53 @@ def initialize
...
         end
 
-        flush(batches)
+        methods.each do |method, options|
+          job_class = method.classify.constantize
+          max_records_per_call = job_class.get_sidekiq_options["max_records_per_call"] || Sidekiq::Grouping::Config.max_records_per_call
+          max_calls_per_minute = job_class.get_sidekiq_options["max_calls_per_minute"] || Sidekiq::Grouping::Config.max_calls_per_minute
+
+          number_of_calls = options[:records] / max_records_per_call
+          number_of_calls += 1 if options[:records] % max_records_per_call > 0
```
Comment (dandemeyere) [inline lib/sidekiq/grouping/actor.rb]: Nice use of mod here to determine if record / max_records has a remainder and needs to be called again to finish off the remainder batch.

**Comment by dandemeyere** [comment]
@florrain great idea to not re-invent the wheel and to utilize an existing system. We have a unique edge case (if the number of records for a batch reaches the max batch number and results in flushes faster than the batch interval), but have you considered requesting a merge to the repo's owners to provide this functionality? You might have to make some abstraction changes to keep it non-specific to thredUP, but it would be great for you and your resume and we wouldn't have to worry about upstream gem updates of our forked version. 

Either way, what's your plan for testing this? Maybe do a lightweight Sidekiq job that simply writes to a log with a timestamp to verify the batches are flushing as expected?

File: lib/sidekiq/grouping/actor.rb (hunk 1 above)
Comment (florrain) [inline lib/sidekiq/grouping/actor.rb:14]: The class Actor is a Celluloid actor or cell that lives in its own thread. Here it's used to read for batches to flush every x seconds. 

I added the `enabled` option to the GEM configuration to disable the event to avoid calling Responsys in environments we don't want to (tests + staging servers). This is what this line is for.

I just realized that it could even be done before the actor gets registered to Sidekiq in `link_to_sidekiq-manager`. I'll move the line down there.

**Comment by florrain** [comment]
In the situation you're pointing, the GEM will make sure it picks `max_number_of_records*max_number_of_calls` jobs out of the batches for that method per minute. Then that number will be split over all the queues of that same worker. So Responsys will never be callsed more than it should be. Well that's how I thought the process in the big code block in `Actor`. A method I want to explode to write tests on.

``` ruby
        methods.each do |method, options|
          job_class = method.classify.constantize
          max_records_per_call = job_class.get_sidekiq_options["max_records_per_call"] || Sidekiq::Grouping::Config.max_records_per_call
          max_calls_per_minute = job_class.get_sidekiq_options["max_calls_per_minute"] || Sidekiq::Grouping::Config.max_calls_per_minute

          number_of_calls = options[:records] / max_records_per_call
          number_of_calls += 1 if options[:records] % max_records_per_call > 0

          if number_of_calls > max_calls_per_minute
            number_of_records_to_process = max_calls_per_minute * max_records_per_call
          else
            number_of_records_to_process = options[:records]
          end

          records_per_queue = number_of_records_to_process / options[:batches].length

          methods[method][:flushable_records] = records_per_queue
        end

        flush(methods)
```

About merging the functionalities, I'm not sure. I changed quite a bit of code and tried to move the complexity into the GEM. But that is something I can work on, I agree it would make it good for everybody.

I tested the changes when I tested our app's PR changes but not at a high load. I tested first with empty jobs and with logs to output the argument passed to the batched job and then with actual Responsys data. 

I want to add specs (on the calculations above) first to ensure my math thoughts make sense and then will do a real test to ensure it handles the throughput correctly.

What I'm more worried about is that the time for the next batch flush is updated when the GEM pushes the big job to Sidekiq and actually not when the job is finished. So in the case where the job's Responsys calls are slow, the next flush will not wait for the first to end before new calls will be triggered. That's something else the GEM could handle (like with a success callback).

Thanks for the review :+1:

**Comment by florrain** [comment]
Closing in favor of https://github.com/thredup/sidekiq-grouping/pull/2
